### PR TITLE
Add Compat entries for #11140 & #11241

### PR DESCRIPTION
**Pull request by ScottPJones**
This adds Compat entries for is_valid_utf32, and all of the isvalid changes.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -298,3 +298,21 @@ end
...
 @test fma(3,4,5) == 3*4+5 == muladd(3,4,5)
+
+# is_valid_utf32
+@test is_valid_utf32(UInt32[65,66,67,0])
+@test !is_valid_utf32(UInt32[65,66,0x110000,0])
+
+# isvalid
+let s = "abcdef", u8 = "abcdef\uff", u16 = utf16(u8), u32 = utf32(u8)
+@test !isvalid(UTF32String(UInt32[65,66,0x110000,0])
+@test !isvalid(Char(0x110000))
```
Comment (tkelman) [inline test/runtests.jl]: this test needs to be something like `@test !isvalid(@compat Char(0x110000))` for 0.3

**Comment by ScottPJones** [comment]
Thanks, I've been trying to figure out how to make it work in 0.3

**Comment by ScottPJones** [comment]
Should be fixed now!

**Comment by tkelman** [comment]
Cool, I think this looks right. Thank you very much!

**Comment by ScottPJones** [comment]
@tkelman Can this be merged now?  (so nobody has issues going back and forth between 0.4 and 0.3.x?)  Thanks!

**Comment by tkelman** [comment]
Sure.

**Comment by hayd** [comment]
Not sure this was quite right as it failed travis https://travis-ci.org/JuliaLang/Compat.jl/jobs/64189228

    MethodError: `isvalid` has no method matching isvalid(::Type{UTF16String}, ::UTF16String)

Does isvalid etc need to be exported outside of the if-block?

**Comment by tkelman** [comment]
Hm, that failed on 0.4.0-dev+5023, so it's actually that the test is trying to use methods that don't exist on Julia master
```
INFO: Testing Compat
WARNING: is_valid_utf32(str::UTF32String) is deprecated, use isvalid(str) instead.
 in depwarn at ./deprecated.jl:62
 in is_valid_utf32 at deprecated.jl:49
 in anonymous at test.jl:91
 in do_test at test.jl:49
 in include at ./boot.jl:252
 in include_from_node1 at loading.jl:133
 in process_options at ./client.jl:320
 in _start at ./client.jl:412
while loading /home/travis/.julia/v0.4/Compat/test/runtests.jl, in expression starting on line 304
ERROR: LoadError: test error in expression: isvalid(UTF16String,u16)
MethodError: `isvalid` has no method matching isvalid(::Type{UTF16String}, ::UTF16String)
Closest candidates are:
  isvalid(::Type{UTF16String}, ::AbstractArray{UInt16,N})
 in anonymous at test.jl:91
 in do_test at test.jl:49
 in anonymous at no file:320
 in include at ./boot.jl:252
 in include_from_node1 at loading.jl:133
 in process_options at ./client.jl:320
 in _start at ./client.jl:412
while loading /home/travis/.julia/v0.4/Compat/test/runtests.jl, in expression starting on line 309
```

**Comment by hayd** [comment]
That was the other possibility! It's annoying that this version number is not printed in the travis log, or available more cleanly from the commit without building (perhaps I'm missing a trick here?).

**Comment by tkelman** [comment]
I specifically made sure that the version number would be printed in the Travis log when we were implementing `language: julia` support, see https://travis-ci.org/JuliaLang/Compat.jl/jobs/64189228#L87

**Comment by hayd** [comment]
@tkelman sorry! I meant on the JuliaLang/julia travis log!

**Comment by tkelman** [comment]
I added a `versioninfo()` line there too: https://github.com/JuliaLang/julia/blob/b4f884d5a28579e926c97ddcb62978d1f1159217/.travis.yml#L62

**Comment by hayd** [comment]
Hmmmmmmmmm, weirdly versioninfo does show the version number locally e.g. +5016... But on travis it just has the branch name:
```
$ /tmp/julia/bin/julia-debug -e 'versioninfo()'
Julia Version 0.4.0-dev
...
```
e.g. look for "versioninfo" on [this build](https://travis-ci.org/JuliaLang/julia/jobs/64357193).

**Comment by tkelman** [comment]
Oh, that is odd. I think what might be going on there is the way travis clones from github is confusing our version_git script, note the warnings a little earlier
```
fatal: No names found, cannot describe anything.
fatal: bad revision '^'
```

**Comment by ScottPJones** [comment]
Did I do something wrong? :sad:

**Comment by tkelman** [comment]
Will need to fix the method error from https://github.com/JuliaLang/Compat.jl/pull/90#issuecomment-106163665 - not sure why Travis didn't catch this the first time around

**Comment by hayd** [comment]
Travis didn't catch it cos it was `Julia Version 0.4.0-dev+4850`.

**Comment by ScottPJones** [comment]
OK, I figured it out... my tests were on v0.3.7, v0.3.8 and v0.4 before isvalid... and this breaks now that isvalid has been added to master, and the Compat code isn't being used, because although the Compat code handled the two argument case, the code in master missed a method for the 2-argument form...
I'm testing the fix now, but it will need to be merged into julia/master (and who knows how long that will take!)  The problem is only in the test though, not in the additions to Compat.jl themselves.
That does show an issue with Compat... how do you know it gives you anything at all like actual implementation of something in the later versions of julia?  You just have to be very careful... but there are no cross-version tests...

**Comment by ScottPJones** [comment]
Should be fixed by #11482 when/if that is merged...

**Pull request merged**